agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,481 @@
+	// cannot find nvidia-smi, the --nv fails to make it available inside the container, e.g.,
+	// env -i /usr/bin/singularity run --nv \\
+	//   docker://determinedai/environments:cuda-11.3-pytorch-1.10-tf-2.8-gpu-24586f0 nvidia-smi
+	// cmd.Env = append(cmd.Env, fmt.Sprintf("PATH=%s", os.Getenv("PATH")))
+
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("starting podman container: %w", err)
+	}
+
+	cont.PID = cmd.Process.Pid
+	cont.Proc = cmd.Process
+	cont.TmpDir = tmpdir
+	cont.Started.Store(true)
+	at := time.Now().String()
+	s.log.Infof("started container %s with pid %d", id, cont.PID)
+
+	return &docker.Container{
+		ContainerInfo: types.ContainerJSON{
+			ContainerJSONBase: &types.ContainerJSONBase{
+				ID:      strconv.Itoa(cont.Proc.Pid),
+				Created: at,
+				Path:    podmanWrapperEntrypoint,
+				Args:    req.ContainerConfig.Cmd,
+				State: &types.ContainerState{
+					Status:    "running",
+					Running:   true,
+					Pid:       cont.Proc.Pid,
+					StartedAt: at,
+				},
+				Image: image,
+				HostConfig: &dcontainer.HostConfig{
+					NetworkMode: req.HostConfig.NetworkMode,
+				},
+			},
+			Config: &dcontainer.Config{
+				ExposedPorts: req.ContainerConfig.ExposedPorts,
+			},
+		},
+		ContainerWaiter: s.waitOnContainer(cproto.ID(id), cont, p),
+	}, nil
+}
+
+func capabilitiesToPodmanArgs(req cproto.RunSpec, args []string) []string {
+	for _, cap := range req.HostConfig.CapAdd {
+		args = append(args, "--cap-add", cap)
+	}
+	for _, cap := range req.HostConfig.CapDrop {
+		args = append(args, "--cap-drop", cap)
+	}
+	return args
+}
+
+func hostMountsToPodmanArgs(m mount.Mount, args []string) []string {
+	var mountOptions []string
+	if m.ReadOnly {
+		mountOptions = append(mountOptions, "ro")
+	}
+	if m.BindOptions != nil && string(m.BindOptions.Propagation) != "" {
+		mountOptions = append(mountOptions, string(m.BindOptions.Propagation))
+	}
+	var options string
+	if len(mountOptions) > 0 {
+		options = fmt.Sprintf(":%s", strings.Join(mountOptions, ","))
+	}
+	return append(args, "--volume", fmt.Sprintf("%s:%s%s", m.Source, m.Target, options))
+}
+
+// ReattachContainer implements container.ContainerRuntime.
+// TODO(DET-9082): Ensure orphaned processes are cleaned up on reattach.
+func (s *PodmanClient) ReattachContainer(
+	ctx context.Context,
+	reattachID cproto.ID,
+) (*docker.Container, *aproto.ExitCode, error) {
+	return nil, nil, container.ErrMissing
+}
+
+// RemoveContainer implements container.ContainerRuntime.
+func (s *PodmanClient) RemoveContainer(ctx context.Context, id string, force bool) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	cont, ok := s.containers[cproto.ID(id)]
+	if !ok {
+		return container.ErrMissing
+	}
+
+	if cont.Started.Load() {
+		return cont.Proc.Kill()
+	}
+	return fmt.Errorf("cannot kill container %s that is not started", id)
+}
+
+// SignalContainer implements container.ContainerRuntime.
+func (s *PodmanClient) SignalContainer(
```

Comment:
Who calls this?

Reply:
Methods in agent.go & container.go. The code here is replicated from singularity.go.